### ISSUE TITLE
pages/ and server/routes/ handling CommonJS/UMD differently

### ISSUE THREAD

**Issue by tommie** [nuxt/nuxt#22126]
### Environment

N/A, since stackblitz can reproduce it.

### Reproduction

https://stackblitz.com/edit/github-ccmqix?file=server%2Froutes%2Fsomeroute.ts

### Describe the bug

I use [`ts-proto`](https://github.com/stephenh/ts-proto) to generate grpc-web (and Protobuf) bindings. It generates code that imports [`@improbable-eng/grpc-web`](https://github.com/improbable-eng/grpc-web). Node fails to import this:

```
  import { grpc } from 'node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.js';
  ^^^^
  SyntaxError: Named export 'grpc' not found. The requested module 'node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.js' is a CommonJS module, which may not support all module.exports as named exports.
  CommonJS modules can always be imported via the default export, for example using:
```

This goes away if enabling transpilation in `nuxt.config.ts`. However, with that enabled, now the client side fails (in browser console):

```
Uncaught SyntaxError: The requested module '/_nuxt/node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.umd.js?v=aa206f89' does not provide an export named 'grpc' (at someproto.ts:1:10)
```

I've tried a couple of changes to the import (see `someproto.ts`), but each one causes errors in various ways.

I'm confused about the fact that "server code" seems to have two different flavors: pages running under SSR and what runs from `server/`. I saw none of this while I was just using `pages/` in SSR, without `server/`.

I suspect they're either using different combinations of CommonJS/UMD import flavors, or there is some difference in TS compilation.

### Additional context

The `grpc-web` module is TS compiled, there is no default export and everything is wrapped in the `grpc` namespace: https://github.com/improbable-eng/grpc-web/blob/master/client/grpc-web/src/index.ts

Note that the `grpc-web` package has two compiled files: a main:CommonJS and a browser:UMD (https://github.com/improbable-eng/grpc-web/blob/master/client/grpc-web/webpack.config.js)

**Comment by tommie** [comment]
Adding for completeness about `grpc-web` specifically:

Running a simple ESM file directly in Node v19 results in the same error:

```js
import { grpc } from "@improbable-eng/grpc-web";
```

However, this is how the grpc-web project itself imports the file in TS: https://github.com/improbable-eng/grpc-web/blob/master/integration_test/ts/src/client.spec.ts

Importing the default works:

```js
import grpc from "@improbable-eng/grpc-web";
```

While in JS (presumably NodeJS), they use `require(...).grpc`: https://github.com/improbable-eng/grpc-web/blob/master/integration_test/ts/_proto/improbable/grpcweb/test/test_pb_service.js

They are certainly doing something fishy and relying on TS compilation, but it also seems to be a fairly vanilly TS and webpack project, so I'd expect it to either always work in Nuxt, or never. Right now it depends on where it's imported.

**Comment by danielroe** [comment]
The pages folder is processed by vite; the server directory is only processed by nitro. Both use rollup under the hood, with different configurations.

The root issue is upstream and should be reported to `grpc-web` - it's invalid Node ESM and can only be consumed by a bundler.

I couldn't reproduce the issue in your stackblitz but would expect something like the below to work:

```ts
// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
  nitro: {
    externals: {
      inline: ["@improbable-eng/grpc-web"]
    }
  },
})
```

**Comment by tommie** [comment]
Thanks for looking into it.

> The root issue is upstream and should be reported to

Yepp, makes sense. Funny thing is they had an issue open about it in 2019 but it was claimed to have resolved itself somehow. I've opened a new issue on grpc-web: https://github.com/improbable-eng/grpc-web/issues/1171

Making a webpack ESM library does solve the issue: https://github.com/improbable-eng/grpc-web/compare/master...tommie:grpc-web:esm?expand=1

I would ask if there's an FAQ to add this to, but I guess this issue itself is that now. :)

> I couldn't reproduce the issue in your stackblitz

Ah, I've changed `someproto.ts` back to the bad case. Sorry about that.

> but would expect something like the below to work:

Played with inline, and that works as a temporary solution. Thanks!
